### PR TITLE
First support for ICESat GLAH06

### PR DESCRIPTION
**Pull request by alex-s-gardner**
@evetion 

Here is the first crack at a GLAH06 reader. Two things you might be able to help with:

1. What is your approach to making nice ascii tables for the function help. I found formatting by hand tricky and didn't manage to get it all

2. ICESat elevations need to be converted from the TOPEX/POSEIDON to WGS84 the ellipsoid. Do you know of a good way to do this in Julia?

### CODE REVIEW

**Comment by alex-s-gardner** [comment]
@evetion

Once you look this over it should be good to merge

File: src/ICESat/GLAH06.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        datetime = unix2datetime.(datetime .+ j2000_offset)
+
+        # convert from TOPEX/POSEIDON to WGS84 ellipsoid using Proj.jl
+        # This pipeline was validated against MATLAB's geodetic2ecef -> ecef2geodetic
+        pipe = Proj.proj_create("+proj=pipeline +step +proj=unitconvert +xy_in=deg +z_in=m +xy_out=rad +z_out=m +step +inv +proj=longlat +a=6378136.3 +rf=298.257 +e=0.08181922146 +step +proj=cart +a=6378136.3 +rf=298.257 +step +inv +proj=cart +ellps=WGS84 +step +proj=unitconvert +xy_in=rad +z_in=m +xy_out=deg +z_out=m +step +proj=axisswap +order=2,1")
+        
+        # the values passed to proj_trans() respect the axis order and axis unit of the official definition ( so for example, for EPSG:4326, with latitude first and longitude next, in degrees)
+        _, _, height_ref = Proj.proj_trans(pipe, Proj.PJ_FWD, (latitude,longitude, height_ref))
+
+        longitude, latitude, height = Proj.proj_trans(pipe, Proj.PJ_FWD, (latitude, longitude, height))
```
Comment (evetion) [inline src/ICESat/GLAH06.jl:55]: Duplication of line 53

File: src/ICESat/GLAH06.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        gt = (
+            longitude = longitude,
+            latitude = latitude,
+            height = height,
+            datetime = datetime,
+            # quality defined according to Smith et al, 2020. DOI: 10.1126/science.aaz5845
```
Comment (evetion) [inline src/ICESat/GLAH06.jl:62]: Might be good to cite it in the docstring as [footnote](https://docs.julialang.org/en/v1/stdlib/Markdown/#Footnote-references).

File: src/ICESat/GLAH06.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+| `longitude`        | `Data_40HZ/Geolocation/d_lon`            | Longitude of segment center, WGS84, East=+            | decimal degrees         |
+| `latitude`         | `Data_40HZ/Geolocation/d_lat`            | Latitude of segment center, WGS84, North=+            | decimal degrees         |
+| `height`           | `Data_40HZ/Elevation_Surfaces/d_elev`    | + `Data_40HZ/Elevation_Corrections/d_satElevCorr`     | m above WGS84 ellipsoid |
+| `datetime`         | `Data_40HZ/DS_UTCTime_40`                | Precise time of aquisiton                             | date-time               |
+| `quality`          | `Data_40HZ/Quality/elev_use_flg`         | & `Data_40HZ/Quality/sigma_att_flg` = 0               |                         |
+|                    | & `Data_40HZ/Waveform/i_numPk` = 1       | & `Data_40HZ.Elevation_Corrections/d_satElevCorr` < 3 | 1 = high quality        |
+| `height_reference` | `land_ice_segments/dem/dem_h`            | Height of the (best available) DEM                    | height above WGS84      |
+
+You can combine the output in a `DataFrame` with `reduce(vcat, DataFrame.(points(g)))`.
```
Comment (evetion) [inline src/ICESat/GLAH06.jl:17]:
```suggestion
You can get the output in a `DataFrame` with `DataFrame(points(g)`.
```
The vcat reduction is required for the multiple beam/groups of ICESat-2.

**Review by evetion — COMMENTED** [review]
Can you also write an integration test? See `test/runtest.jl` for the current suite. Would be good to test that the `points` function works and returns some actual rows, you can copy the `GLAH14` test.

To make the reference file available, add a line like

download_artifact(v"0.1", "GLAH06_634_2131_002_0084_4_01_0001.H5")

File: src/ICESat/GLAH06.jl [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        height_ref[height_ref .== icesat_fill] .= NaN
+
+        datetime = unix2datetime.(datetime .+ j2000_offset)
+
+        # convert from TOPEX/POSEIDON to WGS84 ellipsoid using Proj.jl
+        # This pipeline was validated against MATLAB's geodetic2ecef -> ecef2geodetic
+        pipe = Proj.proj_create("+proj=pipeline +step +proj=unitconvert +xy_in=deg +z_in=m +xy_out=rad +z_out=m +step +inv +proj=longlat +a=6378136.3 +rf=298.257 +e=0.08181922146 +step +proj=cart +a=6378136.3 +rf=298.257 +step +inv +proj=cart +ellps=WGS84 +step +proj=unitconvert +xy_in=rad +z_in=m +xy_out=deg +z_out=m +step +proj=axisswap +order=2,1")
+        
+        # the values passed to proj_trans() respect the axis order and axis unit of the official definition ( so for example, for EPSG:4326, with latitude first and longitude next, in degrees)
+        _, _, height_ref = Proj.proj_trans(pipe, Proj.PJ_FWD, (latitude,longitude, height_ref))
```
Comment (evetion) [inline src/ICESat/GLAH06.jl:53]: This doesn't work on arrays right? You need a broadcast and a zip here? Like: ```Proj.proj_trans.(pipe, Proj.PJ_FWD, zip(latitude,longitude, height_ref))```

File: src/ICESat/GLAH06.jl (hunk 1 above)
Comment (alex-s-gardner) [reply src/ICESat/GLAH06.jl:55]: height_ref != height

File: src/ICESat/GLAH06.jl (hunk 2 above)
Comment (alex-s-gardner) [reply src/ICESat/GLAH06.jl:62]: updated (hopefully I did this correctly)

File: src/ICESat/GLAH06.jl (hunk 3 above)
Comment (alex-s-gardner) [reply src/ICESat/GLAH06.jl:17]: Good catch, thanks!

File: src/ICESat/GLAH06.jl (hunk 4 above)
Comment (alex-s-gardner) [reply src/ICESat/GLAH06.jl:53]: > Can you also write an integration test? See `test/runtest.jl` for the current suite. Would be good to test that the `points` function works and returns some actual rows, you can copy the `GLAH14` test.

Now implemented

**Comment by alex-s-gardner** [comment]
all requested fixes and changes implemented

**Comment by evetion** [comment]
I've fixed some small things and removed the following lines

```julia
latitude = [x[1] for x in pts]::Vector{Float64}
latitude = [x[1] for x in pts]::Vector{Float64}
```

as they should be unchanged from their input (only vertical transformation) and allocate extra memory. For the testfile, it seems all shots have quality=false. We might be just unlucky with the granule, but it's something you might check with other granules.

Otherwise I'm happy to merge this.

**Comment by alex-s-gardner** [comment]
Thanks for the fixes...

I think there are 1.0e-06 differences in latitude because of the slight differences in ellipsoid so it might be best to keep, I'll add back in.

`latitude = [x[1] for x in pts]::Vector{Float64}`

Let be check on the quality masking

**Comment by alex-s-gardner** [comment]
Broadcast was missing in quality flag:
`(i_numPk .== 1) .&`

**Comment by evetion** [comment]
Good catch in the quality broadcast, missed that completely. 

Do you expect differences in the latitude for an ellipsoid transformation? What we pass in are float32s, so they're only accurate up to 7 decimals which would be errors of centimeters or less.

**Comment by alex-s-gardner** [comment]
🤞 we're good to merge

**Comment by alex-s-gardner** [comment]
Yes, I expect very small differences in the latitude for an ellipsoid transformation (at least I think I have that right).. longitude is unchanged. 

But I don't think it matters at all for ICESat data so I'll remove that again

**Comment by alex-s-gardner** [comment]
`latitude = [x[1] for x in pts]::Vector{Float64}` is now commented out

**Comment by alex-s-gardner** [comment]
@evetion you should be good to merge.. then I can start on other items

**Pull request merged**